contracts/bond/Bond.sol

```diff
@@ -120,6 +121,13 @@ contract Bond is Context, ERC20, Ownable, Pausable {
 
         _transfer(address(this), sender, amount);
         emit DebtIssue(sender, symbol(), amount);
+
+        if (balanceOf(address(this)) == 0) {
```

Comment:
question: why is `balance == 0` used as the check for whether the bonding limit has been reached? I think it'd be easier to read if the debt token is explicitly specified for the balance check. The way it is, it implies the balance of remaining debt tokens is checked but I'm not sure.

Reply:
I'm not sure how good the Solidity compiler is 🤔 

The code would change to 
```
uint256 debtTokenBalance = balanceOf(address(this));
if(debtTokenBalance == 0)
```

I'll run a gas comparison , if it's negligible, then I'll refactor 👍

Reply:
Without additional assignment for readability
```
  Contract call:       Bond#deposit
  Transaction:         0x6147631a32d963fb9519bf71146cfd67b12beab233a0eae213c88322bade06e5
  From:                0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc
  To:                  0xcafac3dd18ac6c6e92c921884f9e4176737c052c
  Value:               0 ETH
  Gas used:            137405 of 29021464
  Block #6:            0x913b653ae721d792dad0fb3f1f08d385f4789a8549638d8e9ec15ae47db9d5c8
```

With
```
  Contract call:       Bond#deposit
  Transaction:         0x349de6ce55d21c12d8cc2a0b6edeaf49800a21c250e5e0a7fdfef5bde53b717d
  From:                0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc
  To:                  0xcafac3dd18ac6c6e92c921884f9e4176737c052c
  Value:               0 ETH
  Gas used:            137418 of 29021464
  Block #6:            0x38433cb9b163c5a401844a3ffbdcf94f7e91d32afffa08bea3a15540f45a9c0b
``` 

`137405` vs `137418` == increase of 13 gwei

**observation:** Solidity compiler does not inline the operation, it must be leaving as an additional assignment.

**thought**:  Does compiling with the optimiser make a difference?

With optimiser, without additional assignment
```
  Contract call:       Bond#deposit
  Transaction:         0x0b26016c09fd0fcf669ae78212b524d6969ff2690d6d58a8f24b484b48653ce3
  From:                0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc
  To:                  0xcafac3dd18ac6c6e92c921884f9e4176737c052c
  Value:               0 ETH
  Gas used:            133912 of 29021464
  Block #6:            0x7bdd6a0418d05434d2ee9b53b0fd26f59f3a9512e28b187fd72f5b7a96e481f0
```

With optimiser and additional assignment for readability
```
  Contract call:       Bond#deposit
  Transaction:         0x8a5f40bf9da840d77e117edbeba74a2a2ef971941b17012c7b38f64e19d23ccf
  From:                0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc
  To:                  0xcafac3dd18ac6c6e92c921884f9e4176737c052c
  Value:               0 ETH
  Gas used:            133928 of 29021464
  Block #6:            0xfdbc87e3b072cba1d05f24b4c78006a7e4d059dcd791c9be24bc19dbd1f589bb
```

`133912` vs `133928` == increase of 16 gwei

**observation**:  
1) using the Solidity optimiser notably reduces gas usage. 
2) additional assignments for greater in source readability has a cost in bytecode efficiency.

Reply:
Of course ...private helper methods!

The optimizer will line certain functions https://docs.soliditylang.org/en/v0.8.9/internals/optimizer.html

Instead of the assignment, using a helper method that will be inline by the compiler will yield both the improved source readability and gas efficiency.
```
    function _debtTokenBalance() private view returns(uint256) {
        return balanceOf(address(this));
    }
```


```
  Contract call:       Bond#deposit
  Transaction:         0x0b26016c09fd0fcf669ae78212b524d6969ff2690d6d58a8f24b484b48653ce3
  From:                0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc
  To:                  0xcafac3dd18ac6c6e92c921884f9e4176737c052c
  Value:               0 ETH
  Gas used:            133912 of 29021464
  Block #6:            0x7e608483d91a32345f88a3c9511049910b46de7e2932198b42995c293eae3d49
```

`133912` vs `133912` == 0 gwei differennce

**observation** : improve code readability with private helper methods!

Reply:
To be clearer, the comparison to zero is based on the expectation that the balance referenced by `this` address is the balance of debt remaining, which is different than an actual balance of value. Is that right? If so, I'd suggest inverting it so that it 's based on value contained. 

Separately, I believe the `constant` keyword will result in inlining.

Reply:
>  Is that right?

🤔 not sure precisely on that statement, but I'll re-phrase as the check is whether the balance of debt tokens held by the contract (the bond is met, only when all debt tokens are issued).

> Separately, I believe the constant keyword will result in inlining.

Yes ...but the balance retrieval cannot be a Solidity constant 😞